### PR TITLE
MIR required_consts, mentioned_items: ensure we do not forget to fill these lists

### PR DESCRIPTION
**Pull request by RalfJung**
Bodies initially get created with empty required_consts and mentioned_items, but at some point those should be filled. Make sure we notice when that is forgotten.

### CODE REVIEW

**Comment by RalfJung** [comment]
This currently ICEs for async closures, looks like we do indeed forget to compute these lists there.

@compiler-errors do async closures not go through the normal MIR passes? In that case they have to manually invoke the right passes to compute required_consts and mentioned_items. Where should that be done?

**Comment by RalfJung** [comment]
Async drop also doesn't seem to compute the required_consts.
Cc @azhogin @zetanumbers

**Review by compiler-errors — APPROVED** [review]
Good change 👍

**Comment by compiler-errors** [comment]
Oops, didn't see this was still failing.

@bors r-

**Comment by compiler-errors** [comment]
> @compiler-errors do async closures not go through the normal MIR passes?

They should go through normal MIR passes. We copy the body here:

https://github.com/rust-lang/rust/blob/c0e32983f5b06a6f7d8cc776ccac71de6512ed6d/compiler/rustc_mir_transform/src/coroutine/by_move_body.rs#L199

Then on subsequent passes, we apply the pass to the `by_move_body` if it gets applied to the parent body:

https://github.com/rust-lang/rust/blob/97ac52f579fe1003a162324d448dad43a942b5f5/compiler/rustc_mir_transform/src/pass_manager.rs#L186-L190

Are we not populating `required_consts` and `mentioned_items` with a pass via the pass manager?

**Comment by RalfJung** [comment]
No those are not normal passes. `required_consts` usually gets populated very early, in `fn mir_promoted`. `mentioned_items` gets populated in `fn inner_optimized_mir`.

`mentioned_items` used to be a normal pass but then it doesn't get run on custom MIR that is "injected" after this pass, which is bad...

**Comment by RalfJung** [comment]
> So any passes previously applied should still be okay. Then on subsequent passes, we apply the pass to the by_move_body if it gets applied to the parent body:

Wait so when passes are applied to one body we actually apply them to another? What a glorious hack.^^

**Comment by compiler-errors** [comment]
Hm. So I guess we could just manually handle the closure "by move" body manually here, but I fear that we'll just forget to apple some other *special* pass in the future just like these two.

> Wait so when passes are applied to one body we actually apply them to another? What a glorious hack.^^ But super hard to follow...

It's the only way that I saw for us to ensure that we're applying all the same passes to the original async closure and the "by move" body in lockstep, since the latter is generated on the fly really early during `mir_build`.

**Comment by RalfJung** [comment]
Not even required_consts are set which are usually set *very* early in MIR construction, so it also seems like this doesn't go through the normal MIR builder, or so?

The by-move body should just be its proper independent own MIR, and then when a query asks for that MIR it should go through the pass manager itself. Why should it "ride along" the other body?

**Comment by compiler-errors** [comment]
Perhaps I should just bite the bullet and synthesize real def ids for async closure "by move" bodies so I can just feed `mir_built` and not worry about all of this 🤔

I guess you can probably do whatever to unblock this, or if you have better ideas I'd love to hear them.

**Comment by RalfJung** [comment]
> Perhaps I should just bite the bullet and synthesize real def ids for async closure "by move" bodies 

Yes, I think that's what I meant. Having one DefId refer to two bodies seems like a recipe for confusion. It's already bad enough with const MIR and runtime MIR, let's not add another dimension.^^

**Comment by compiler-errors** [comment]
> The by-move body should just be its proper independent own MIR, and then when a query asks for that MIR it should go through the pass manager itself. Why should it "ride along" the other body?

Because it's a lot of feeding and I was mostly following the strategy that the drop shim does :D

**Comment by RalfJung** [comment]
Drop shim has a proper DefId though... it does have its own pass manager, which is a bit annoying (and which this PR has to touch), but AFAIK that's mostly because we don't want to run the full set of passes on drop shims.

**Comment by RalfJung** [comment]
Though *some* drop shims also seem to bypass that pass manager, not sure what's going on there...
```
  thread 'rustc' panicked at compiler/rustc_monomorphize/src/collector.rs:1232:26:
  required_consts for DefId(2:2672 ~ core[1802]::ptr::drop_in_place) have not yet been set
```

**Comment by compiler-errors** [comment]
It's arguable to say that drop shim has a proper def id. It uses the def id of `drop_in_place`, which is definitely wrong because the shim is semi-substituted with a type (stored in the `DropShim`) and therefore is parameterized over a random set of generics that's almost certainly not the `<T>` that the function is generic over. I think you probably already know we have lots of hacks for handling drop shims, lol.

But that's kinda unrelated. I'll look into getting this fixed the right way, but it may be much harder than it seems, so it's up to you if you want to block this PR on that work.

**Comment by compiler-errors** [comment]
What's luckily different from the drop shim is that the by-move body doesn't need to be generated after the coroutine transformation pass, so we may be lucky and it turns out much simpler than expected :D

**Comment by RalfJung** [comment]
The first problem is earlier -- the body we see and clone in ByMoveBody already doesnt have its required_consts set. So something is already going wrong before that.

**Comment by RalfJung** [comment]
Oh wait, is mir_built before mir_promoted? In that case that's expected, I guess.

Even the pass manager for normal MIR is already such a mess, it's near impossible to remember what happens in which order.^^

**Comment by RalfJung** [comment]
Let's see, hopefully this is enough.

**Comment by RalfJung** [comment]
> It's arguable to say that drop shim has a proper def id. It uses the def id of `drop_in_place`, which is definitely wrong because the shim is semi-substituted with a type (stored in the `InstanceKind::DropShim`) and therefore is parameterized over a random set of generics that's almost certainly not the `<T>` that the function is generic over. I think you probably already know we have lots of hacks for handling drop shims, lol.
> 
> But that's kinda unrelated. I'll look into getting this fixed the right way, but it may be much harder than it seems, so it's up to you if you want to block this PR on that work.

It's magic, but knowing that it is `drop_in_place` plus knowing the type is enough to uniquely identify the MIR... hm but I guess that's an `Instance`, not an `InstanceKind`, and usually `Body` is for an `InstanceKind`... so okay I agree, that's also a hack.^^

>  is parameterized over a random set of generics that's almost certainly not the `<T>` that the function is generic over

No idea what you mean here. The `T` in `drop_in_place::<T>` is what it is parameterized over, how is there a "random set of generics"?

**Comment by compiler-errors** [comment]
> No idea what you mean here.

I'm talking about why we have https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/enum.InstanceKind.html#method.has_polymorphic_mir_body, because the body you get back from `InstanceKind::DropShim(_, Some(ty))` is not parameterized over `T`, but whatever generics show up in `ty`. But this convo is kinda unrelated now, so if you want to keep talking about how drop shims are cursed then DM me or start a zulip thread or sth :D

**Comment by RalfJung** [comment]
I think we can agree that drops shims and asyn-by move bodies are both cursed, each in their own unique ways.  :)

I hope that what this PR does now works. It's a hack, but we should notice fairly easily when it goes wrong (that's exactly why the first version of the PR ICEd).

**Comment by compiler-errors** [comment]
Cool. These hacks should become redundant with #128506, but they're totally fine now.

@bors r+
